### ISSUE TITLE
fix: ValidatedNested with transform not validating

### ISSUE THREAD

**Issue by ohad-griiip** [typestack/class-validator#1473]
### _Versions_: 
*  class-transformer: 0.4.0
*  class-validator: 0.13.2

### Description
Using _Transform_ from _class-transformer_ and _ValidateNested_ of _class-validator_ seems to skip validation of nested fields. 

**Minimal code-snippet showcasing the problem**
```ts
import 'reflect-metadata'
import { IsInt, ValidateNested, validateSync } from 'class-validator'
import { plainToClass, Transform, Type } from 'class-transformer'

class Son {
    @IsInt()
    num!: number
}

class Parent {
    @ValidateNested()
    @Type(() => Son)
    @Transform(({ value }) => ({ num: 'a' }))
    son!: Son
}

const parent = plainToClass(Parent, { son: { num: 'abc' } })
console.log('errors: ', validateSync(parent).toString())  // Passing with no validation errors
```

### Expected behavior
Displaying validation error - in this case not passing the _isInt_. 

### Actual behavior
Passing validation

**Comment by ruscon** [comment]
validation and transformations work with class instances.

try this
```typescript
class Parent {
    @ValidateNested()
    @Type(() => Son)
    @Transform(({ value }) => {
      // change and return the current instance of `Son` or create new one
       value. num =  'a';
       return value;
    })
    son!: Son
}
```

**Comment by molaeiali** [comment]
Same issue here:
```
		"class-transformer": "^0.5.1",
		"class-validator": "^0.13.2",
```
I want to be able to have this input in my query params:
:3000?sort={"name":1}&page=1&limit=20

```
class SortDto {
	@IsNumber()
	@IsIn([1, -1])
	readonly name!: number;
}

export class PaginateDto {
	@Transform(({ value }) => JSON.parse(value), { toClassOnly: true })
	@ValidateNested()
	readonly sort!: SortDto;

	@Type(() => Number)
	@IsOptional()
	@IsInt()
	@Min(0,)
	readonly page: number = 0;

	@Type(() => Number)
	@IsOptional()
	@IsInt()
	@Min(0)
	readonly limit: number = 10;
}
```

But when I pass anything as sort (even invalid values like { "name": "A" }, I'm receiving this in my controller:
```
{ page: 1, limit: 20, sort: {} }
```


I need to transform my sort variable from string to JSON as it is being passed through query params

**Comment by molaeiali** [comment]
> validation and transformations work with class instances.
> 
> try this
> 
> ```ts
> class Parent {
>     @ValidateNested()
>     @Type(() => Son)
>     @Transform(({ value }) => {
>       // change and return the current instance of `Son` or create new one
>        value.num = 'a';
>        return value;
>     })
>     son!: Son
> }
> ```

Tried 
```
	@Transform(
		({ value }) => {
			value = JSON.parse(value);
			return value;
		},
		{ toClassOnly: true },
	)
```
No difference

**Comment by manistra** [comment]
For my particular issue I wanted to parse a string into an object and validate it... this works for me: 

```
export class UploadDocumentDto {
  @ValidateNested()
  @Transform(({ value }) => plainToClass(UploadDocument, JSON.parse(value)), {
    toClassOnly: true,
  })
  @Type(() => UploadDocument)
  readonly documents: UploadDocument[];
}
export class UploadDocument {
  @IsNotEmpty()
  @IsString()
  id: string;

  @IsNotEmpty()
  @IsString()
  name: string;
}

```

**Comment by manistra** [comment]
> > validation and transformations work with class instances.
> > try this
> > ```ts
> > class Parent {
> >     @ValidateNested()
> >     @Type(() => Son)
> >     @Transform(({ value }) => {
> >       // change and return the current instance of `Son` or create new one
> >        value.num = 'a';
> >        return value;
> >     })
> >     son!: Son
> > }
> > ```
> 
> Tried
> 
> ```
> 	@Transform(
> 		({ value }) => {
> 			value = JSON.parse(value);
> 			return value;
> 		},
> 		{ toClassOnly: true },
> 	)
> ```
> 
> No difference

Instead of just: 
```return value```
try using:
```return  plainToClass(DesiredClassDto, value)```


For some reason `{ toClassOnly: true }` alone doesn't do the trick, and if I'm not wrong for values to be validated they need to be part of a dto class instance.

**Comment by camargosproj** [comment]
> For my particular issue I wanted to parse a string into an object and validate it... this works for me:
> 
> ```
> export class UploadDocumentDto {
>   @ValidateNested()
>   @Transform(({ value }) => plainToClass(UploadDocument, JSON.parse(value)), {
>     toClassOnly: true,
>   })
>   @Type(() => UploadDocument)
>   readonly documents: UploadDocument[];
> }
> export class UploadDocument {
>   @IsNotEmpty()
>   @IsString()
>   id: string;
> 
>   @IsNotEmpty()
>   @IsString()
>   name: string;
> }
> ```
I had the same issue. This worked for me. 

"class-validator": "^0.13.2" plainToClass is deprecated, so I used plainToInstance 

Thanks a lot!

**Comment by Tushar-zaapi** [comment]
> > For my particular issue I wanted to parse a string into an object and validate it... this works for me:
> > ```
> > export class UploadDocumentDto {
> >   @ValidateNested()
> >   @Transform(({ value }) => plainToClass(UploadDocument, JSON.parse(value)), {
> >     toClassOnly: true,
> >   })
> >   @Type(() => UploadDocument)
> >   readonly documents: UploadDocument[];
> > }
> > export class UploadDocument {
> >   @IsNotEmpty()
> >   @IsString()
> >   id: string;
> > 
> >   @IsNotEmpty()
> >   @IsString()
> >   name: string;
> > }
> > ```
> 
> I had the same issue. This worked for me.
> 
> "class-validator": "^0.13.2" plainToClass is deprecated, so I used plainToInstance
> 
> Thanks a lot!

Yes working for me as well with plainToInstance. Thanks a lot both of you

**Comment by aivazoff** [comment]
> For my particular issue I wanted to parse a string into an object and validate it... this works for me:
> 
> ```
> export class UploadDocumentDto {
>   @ValidateNested()
>   @Transform(({ value }) => plainToClass(UploadDocument, JSON.parse(value)), {
>     toClassOnly: true,
>   })
>   @Type(() => UploadDocument)
>   readonly documents: UploadDocument[];
> }
> export class UploadDocument {
>   @IsNotEmpty()
>   @IsString()
>   id: string;
> 
>   @IsNotEmpty()
>   @IsString()
>   name: string;
> }
> ```

I had the same issue. This worked for me.
Thank you very much!

**Comment by satya-nutella** [comment]
How would this work when I have a discriminated union for the array?

**Comment by vic1707** [comment]
Works fine.
Proper fix would bring support for plain objects which would be better as it allows support for `enableImplicitConversion`.

My current (simplified) use-case is with filters as query params.
I pass in `?filters=price:10~15;age:3|5` and use a custom pipe `ParseFiltersPipe` to convert it to `{ price: '10~15', age: '3|5' }`.
Then on my DTO
```typescript
export class FiltersDto {
  @Expose()
  @Transform(({ value }) => PriceRangeFilter.PARSE(value), { toClassOnly: true })
  @ValidateNested()
  @Type(() => PriceRangeFilter)
  price: PriceRangeFilter;

  @Expose()
  @Transform(({ value }) => ListValuesFilter.PARSE(value), { toClassOnly: true })
  @ValidateNested()
  @Type(() => ListValuesFilter)
  age: ListValuesFilter;
}
```
Each filter being a utility class with its own way of parsing I end up after the transforms with `{price: { min: '10', max: '15'}` and `{age: ['3','5']}}` which would end up being validated and give me my validated `FiltersDto` object

My current workaround involves ReflectMetadata to access the `PARSE()` method in `ParseFiltersPipe`, but I'd prefer a proper fix
